include/sys/devfs.h

```diff
@@ -1,35 +1,73 @@
 #ifndef _SYS_DEVFS_H_
 #define _SYS_DEVFS_H_
 
-#define DEVFS_NAME_MAX 64
+#include <sys/types.h>
+#include <sys/refcnt.h>
 
 typedef struct vnode vnode_t;
 typedef struct devfs_node devfs_node_t;
 typedef struct vnodeops vnodeops_t;
+typedef struct devnode devnode_t;
+typedef struct file file_t;
+typedef struct uio uio_t;
 
-/* If parent is NULL new device will be attached to root devfs directory.
- * If vnode_p is not NULL, on success *vnode_p will point to the vnode of
- * the newly created device. */
+/*
+ * Device file node.
+ */
+
+typedef int (*dev_open_t)(devnode_t *dev, int flags, file_t *fp);
+typedef int (*dev_close_t)(devnode_t *dev, int flags);
+typedef int (*dev_read_t)(devnode_t *dev, uio_t *uio);
+typedef int (*dev_write_t)(devnode_t *dev, uio_t *uio);
+typedef int (*dev_ioctl_t)(devnode_t *dev, u_long cmd, void *data, int flags);
+
+typedef enum {
+  DT_OTHER = 0,    /* other non-seekable device file */
+  DT_SEEKABLE = 1, /* other seekable device file (also a flag) */
+  /* TODO: add DT_CONS (!) and DT_DISK. */
+} dev_type_t;
+
+typedef struct devops {
+  dev_type_t d_type;   /* device type */
+  dev_open_t d_open;   /* prepare device for devfs operations */
+  dev_close_t d_close; /* called when file referring to the device is closed */
+  dev_read_t d_read;   /* read bytes form a device file */
+  dev_write_t d_write; /* write bytes to a device file */
+  dev_ioctl_t d_ioctl; /* read or modify device properties */
+} devops_t;
+
+typedef struct devnode {
+  devops_t *ops;   /* device operation table */
+  void *data;      /* device specific data */
+  size_t size;     /* opened device node size (for seekable devices) */
+  refcnt_t refcnt; /* number of open files referring to this device */
+} devnode_t;
+
+/* If parent is NULL new device will be attached to root devfs directory. */
+int devfs_makedev_new(devfs_node_t *parent, const char *name, devops_t *devops,
+                      void *data, devnode_t **dev_p);
+int devfs_makedir(devfs_node_t *parent, const char *name, devfs_node_t **dir_p);
+
+/* TODO: rewrite all device drivers which use devfs to rely on `devops`
+ * instead `vnodeops`. FTTB obsolete device drivers use the following
+ * function instead `devfs_makedev_new`. */
 int devfs_makedev(devfs_node_t *parent, const char *name, vnodeops_t *vops,
                   void *data, vnode_t **vnode_p);
```

Comment:
That's how it was introduced at first - unfortunately it introduced changes to 10 files. Since our main focus of this PR is device node interface I insist on keeping this PR small.

In fact I don't like our current `devfs_makedev` interface at all. Its' caller must deal explicitly with `devfs_node_t *` and directory structure of `devfs` filesystem. This is simply ugly. Please refer to [make_dev_s](https://www.freebsd.org/cgi/man.cgi?query=make_dev_s) on how it should be organised.